### PR TITLE
[#173150927] AWS redis update manager tool

### PR DESCRIPTION
**Pull request by tlwr**
[Story](https://www.pivotaltracker.com/story/show/173150927)

What
----

Semi-automate the listing, previewing, and emailing of elasticache service updates.

Adds a CLI tool in `tools` called `aws_redis_update_manager` that uses:
- AWS API
- PaaS Accounts API
- CF API

to
- `print` (print which org managers are going to be emailed about which instances)
- `preview` (preview the format of the email which you are going to send)
- `email` (email all of the relevant people)

How to review
-------------

CI

Code review commit by commit

Dry run in your development environment:
- Sign in to CF dev env
- Create a redis instance
- Delete the line `.select { |update| update.update_action_status == 'not-applied' }` from `tools/aws_redis_update_manager/lib/elasticache_update_finder.rb` (so you don't have to wait for an update
- `./tools/aws_redis_update_manager/aws_redis_update_manager` with the relevant CLI flags

See [team manual PR](https://github.com/alphagov/paas-team-manual/pull/351) for guidance

Gotchas
-------

- FNV hash function is implemented in ruby as a util, because the gems I tried didn't work without monkey patching
- Not very ergonomic CLI

Example usage
-------

```
AWS_REGION=eu-west-1 gds aws paas-prod -- \
  ./tools/aws_redis_update_manager/aws_redis_update_manager \
  --action preview \
  --notify-api-key "$( command to get this from credhub )" \
  --paas-accounts-url https://accounts.cloud.service.gov.uk \
  --paas-accounts-username "$( command to get this from credhub )" \
  --paas-accounts-password "$( command to get this from credhub )" \
  --maintenance-window-date 'Tuesday 16 June 2020' \
  --maintenance-window-time-range '0600-0800' \
  --alt-maintenance-window-date 'Thursday 18 June' \
  --alt-maintenance-window-time-range '1800-2000' \
  --region Ireland
```

Who can review
--------------

Not @tlwr

### CODE REVIEW

File: tools/aws_redis_update_manager/lib/tenant_notifier.rb [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  - Click “Explore your costs and usage”
+  - Click the dropdown “Services and apps”
+  - Look for any “tiny” Redis plans without “ha” in the name, e.g. “redis tiny-4.x”
+  - Repeat for any other organisations you manage
+
+  # How to upgrade your Redis tiny plans to highly available plans
+
+  - Follow these instructions for every tiny Redis instance you want to upgrade: https://docs.cloud.service.gov.uk/deploying_services/redis/#upgrade-redis-service-plan
+  - Upgrade to the “ha” version of your plan, e.g. from “redis tiny-4.x” to “redis tiny-ha-4.x”
+
```
[the remark this answers is not in the document]
Comment (tlwr) [reply tools/aws_redis_update_manager/lib/tenant_notifier.rb:56]:
```
cf marketplace -s redis | grep -- -ha | grep tiny
tiny-ha-3.2          1.5GB RAM, highly-available, daily backups                                                         paid
tiny-ha-4.x          1.5GB RAM, highly-available, daily backups                                                         paid
tiny-ha-5.x          1.5GB RAM, highly-available, daily backups                                                         paid

File: tools/aws_redis_update_manager/lib/cloud_foundry_service_finder.rb [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+    resources.uniq { |r| r.fetch('guid') }
+  end
+
+  def paginate_v2(url)
+    path = cf_curl_path(url)
+    resources = []
+
+    loop do
+      body = `cf curl '#{path}'`
```
[the remark this answers is not in the document]
Comment (tlwr) [reply tools/aws_redis_update_manager/lib/cloud_foundry_service_finder.rb:67]: Only gem I could find is archived

File: tools/aws_redis_update_manager/lib/tenant_notifier.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (AP-Hunt) [reply tools/aws_redis_update_manager/lib/tenant_notifier.rb:56]: Whoops! I missed it on the marketplace page.

File: tools/aws_redis_update_manager/lib/cloud_foundry_service_finder.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (AP-Hunt) [reply tools/aws_redis_update_manager/lib/cloud_foundry_service_finder.rb:67]: Aha, OK. All good then.

File: tools/aws_redis_update_manager/lib/tenant_notifier.rb [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  PAAS_SUPPORT_REPLY_TO_ID = '76d63d5f-e140-6a37-92ae-fc0d0d136f6f'
+
+  TEMPLATE = <<~MESSAGE
+  Dear GOV.UK PaaS tenant,
+
+  We are contacting you, as the managers of the <%= org_name %> org, to ask you to review your choice of Redis plans ahead of an upgrade.
+
+  This will help you avoid unnecessary downtime for your users.
+
+  We will be upgrading these Redis instances in two weeks time, on <%= maintenance_window_date %>, between <%= maintenance_window_time_range %>:
```
[the remark this answers is not in the document]
Comment (tlwr) [reply tools/aws_redis_update_manager/lib/tenant_notifier.rb]: Resolved in fa31f0c

File: tools/aws_redis_update_manager/lib/tenant_notifier.rb [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  - <%= si.org_name %> / <%= si.space_name %> / <%= si.instance_name %>
+  <% end %>
+
+  # What this means for you
+
+  If you are using any of our “highly available” Redis plans, your users will only experience a small amount of downtime when the upgrade happens: up to a minute.
+
+  However, if you are using any “tiny” Redis plans that are not “highly available”, your users may experience up to 30 minutes of downtime.
+
+  If you want to minimise this downtime, we recommend that you upgrade these tiny instances to “highly available” plans, ahead of the upgrade on <%= maintenance_window_date %>.
```
[the remark this answers is not in the document]
Comment (tlwr) [reply tools/aws_redis_update_manager/lib/tenant_notifier.rb]: Resolved in fa31f0c

**Pull request merged**